### ISSUE TITLE
Creating a no-op circuit breaker policy

### ISSUE THREAD

**Issue by Okeanos** [jhalterman/failsafe#93]
We are currently evaluating Failsafe for our usecase and would like to test whether it suits our needs in a non intrusive way. For that we reason a noop approach that would simply log all state transitions according to the configuration would work.
Is anything like that supported out of the box or via some trickery concerning transition thresholds? E.g by not setting a withDelay() value or similar.

**Comment by jhalterman** [comment]
You mean you'd like state transitions to occur but for `CircuitBreakerOpenException` to not be thrown? That's currently not possible.

You can of course use the `onOpen`, `onHalfOpen`, `onClose` handlers to do your logging and for the purposes of testing, wrap your executions in something that ignores any `CircuitBreakerOpenException`s that are thrown.

**Comment by Okeanos** [comment]
Hm, yes and no. 

We already have logging enabled in the `on*` methods (which are very handy indeed). We would also like to Failsafe to change state with a given configuration but (essentially) prevent it from staying in the `open` state and thus preventing modifying behavior as the `CircuitBreakerOpenException` is never thrown.
As far as I can currently see the only option to achieve that is by omitting `withDelay()` from the initial configuration. In that case, regardless of whether or not the circuit breaker changes to `open` it should immediately/on the next request change to `half-open` and never throw the `CircuitBreakerOpenException`.
Question is … is that intended, configurable behavior or merely a side effect of something else entirely? Are there any consequences to such a configuration when the application/circuit breaker has to manage a lot of requests?

**Comment by jhalterman** [comment]
> Question is … is that intended, configurable behavior or merely a side effect of something else entirely?

It's somewhat intended. I say somewhat, because a delay is pretty much always needed to indicate when the circuit is safe to test again. That's part of the premise of the pattern - that the delay gives the system under load some time to recover. Defaulting to no delay and not explicitly requiring one may be something we should change (though that's a separate issue).

I don't think there are any consequences to having a delay or not in terms of servicing lots of requests. It's a [simple check](https://github.com/jhalterman/failsafe/blob/master/src/main/java/net/jodah/failsafe/internal/OpenState.java#L31-L34) to determine if the delay has been exceeded. Each state change does result in a [new object](https://github.com/jhalterman/failsafe/blob/master/src/main/java/net/jodah/failsafe/CircuitBreaker.java#L461-L470) being allocated but, that's about it.

**Comment by whiskeysierra** [comment]
> As far as I can currently see the only option to achieve that is by omitting withDelay() from the initial configuration. In that case, regardless of whether or not the circuit breaker changes to open it should immediately/on the next request change to half-open and never throw the CircuitBreakerOpenException.

That is incorrect. It only allows a maximum of concurrent executions in half-open, by default 1. It will throw `CircuitBreakerOpenException`.

**Comment by Okeanos** [comment]
Sadly, I am no longer attached to the project in which we originally used this library so I cannot speak for the current state of things but at the time the `CircuitBreaker` was used in conjunction with `Failsafe`. I have since also worked in a project where the `CircuitBreaker` was used in a standalone mode.
Still, I am certain that the use case is will crop up again (if not for me, then for somebody else). So yes, generally speaking I'd still be interested in that feature albeit without an actual project attached right now.

**Comment by jhalterman** [comment]
Will leave this open for now, but am interested to see if anyone else has this and can describe their use case.
